### PR TITLE
Add append option to add_entries processor

### LINKED ISSUE #4129: Append values to lists in an event
**Is your feature request related to a problem? Please describe.**
Append is a common operation for lists. As a Data Prepper user, I would like to be able to append values to a list in the event. 

**Describe the solution you'd like**
`add_entries` processor can be enhanced to achieve this by adding an `append_if_key_exists` option:
```
    - add_entries:
        entries:
          - key: "fruits"
            value: "banana"
            append_if_key_exists: true
```
Example1:
Input:
```json
{"fruits": "apple"}
```
output:
```json
{"fruits": ["apple", "banana"]}
```

Example2:
Input:
```json
{"fruits": ["apple", "watermelon"]}
```
output:
```json
{"fruits": ["apple", "watermelon", "banana"]}
```

**Describe alternatives you've considered (Optional)**
A separate `list_append` processor. 

**Additional context**
#3967

### PR DESCRIPTION
**Pull request by oeyh**
### Description
Adds an `append_if_key_exists` option to `add_entries` processor as described in #4129 
 
### Issues Resolved
Resolves #4129 
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by graytaylor0** [comment]
Would it make sense to instead have something like this

```
action_when_key_exists: append //  can be one of `append`, `overwrite`, `skip`
```

Not a blocking comment just something to consider

**Review by graytaylor0 — APPROVED** [review]

**Comment by oeyh** [comment]
> action_when_key_exists: append //  can be one of `append`, `overwrite`, `skip`
@graytaylor0 I like it, making the config more concise. I can do the change if other reviewers think the same.

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessorConfig.java [hunk 1]
```diff
@@ -58,19 +61,29 @@ public boolean getOverwriteIfKeyExists() {
...
+        }
+
         public String getAddWhen() { return addWhen; }
 
         @AssertTrue(message = "Either value or format or expression must be specified, and only one of them can be specified")
         public boolean hasValueOrFormatOrExpression() {
             return Stream.of(value, format, valueExpression).filter(n -> n!=null).count() == 1;
         }
 
+        @AssertTrue(message = "overwrite_if_key_exists and append_if_key_exists can not be set at the same time.")
```
Comment (dlvenable) [inline data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessorConfig.java:75]: > overwrite_if_key_exists and append_if_key_exists can not be set at the same time.

Please change to:

> overwrite_if_key_exists and append_if_key_exists can not be set to true at the same time.

They can be "set" together.

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessor.java [hunk 2]
```diff
@@ -97,4 +101,30 @@ public boolean isReadyForShutdown() {
     @Override
     public void shutdown() {
     }
+
+    private void mergeValueToEvent(final Event recordEvent, final String key, final Object value) {
```
Comment (dlvenable) [inline data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessor.java:105]: Consider combining these with the use of Java 8 lambdas.

Here is some starting point for one way to do it.

```
private void mergeValueToEvent(final Event recordEvent, final String key, final Object value)
  mergeValue(recordEvent, () -> recordEvent.getMetadata().getAttribute(key), newValue -> recordEvent.put(key, newValue);
}

mergeValue(Supplier<Object> getter, Consumer<Object> setter) {
          final Object currentValue = getter.get();
        final List<Object> mergedValue = new ArrayList<>();
        if (currentValue instanceof List) {
            mergedValue.addAll((List<Object>) currentValue);
        } else {
            mergedValue.add(currentValue);
        }

        mergedValue.add(value);
        setter.accept(mergedValue);
}
```

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Comment by dlvenable** [comment]
> Would it make sense to instead have something like this
> 
> ```
> action_when_key_exists: append //  can be one of `append`, `overwrite`, `skip`
> ```
> 
> Not a blocking comment just something to consider

@graytaylor0 , I like this idea. But, I also think we should aim for consistency. Right now grok has `overwrite_when_key_exists`. Do you think grok could also move toward using this approach? I'd really like to see all the processors use similar behavior. And I think this enum could facilitate that.

**Review by dlvenable — APPROVED** [review]

**Comment by oeyh** [comment]
These processors have similar `overwrite_if_key_exists` options:
- copy_values
- rename_keys
- parse_json
- parse_ion
- key_value

grok processor has a `keys_to_overwrite` option that determines overwrite or append (default) for each grok generated fields.

I will open a follow-up issue for these and we can discuss further there.

File: data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessor.java (hunk 2 above)
Comment (oeyh) [reply data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/AddEntryProcessor.java:107]: Thanks for the suggestion!

**Review by graytaylor0 — APPROVED** [review]

**Pull request merged**